### ISSUE TITLE
Distributed Data Parallel +cuda.amp shows incorrect unused parameters for lstm with state burn-in

### ISSUE THREAD

**Issue by axeln31459** [pytorch/pytorch#79841]
### 🐛 Describe the bug

If I use DDP and Autocast I get errors about unused parameters (or the list of unused parameters if I set find_unused_parameters=True).
DDP without Amp does not generate these errors.
This only happens for an R2D2-like RL sequence LSTM model (where you sample sequences and a  stale initial hidden state from replay and then burn-in the model with no-grad on the start of the sequence to get a better hidden state that you then use to run the gradients for the later steps in the sequence from the updated hidden state)
Looks like a bug to me ? 
Thanks.

Minimal example to reproduce: change autocast=True to False to test that it works without amp:
```

import torch.nn as nn
import torch
import numpy as np
import torch.multiprocessing as mp
import os
import torch.distributed as dist

class R2D2(nn.Module):
    def __init__(self):
        super(R2D2, self).__init__()
        self.conv=nn.Conv2d(4, 16, kernel_size=3,stride=4)
        self.core=nn.LSTM(input_size=16*21*21,hidden_size=512,batch_first=True)
        self.head = nn.Linear(512,15)


    def forward(self,state,hidden_state):
        B=state.shape[0]
        T=state.shape[1]
        state=state.flatten(start_dim=0, end_dim=1)
        latent=self.conv(state)
        latent=latent.reshape(B,T,-1)
        x,hidden_state=self.core(latent,hidden_state)
        x=self.head(x.flatten(start_dim=0, end_dim=1)).reshape(B,T,-1)
        return x,hidden_state


def run(rank,size):
    autocast=True
    torch.cuda.set_device(rank)
    burnin=40
    model=R2D2().cuda()
    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[rank],
                                                      output_device=rank,find_unused_parameters=True)
    scaler=torch.cuda.amp.GradScaler()
    optimizer = torch.optim.Adam(model.parameters(), 1e-5)
    target=torch.rand(64,80,1, dtype=torch.float32).detach().cuda()
    for i in range(1):
        with torch.cuda.amp.autocast(enabled=autocast):
            optimizer.zero_grad(set_to_none=True)
            state=torch.from_numpy(np.random.randint(0, 255, (64,120, 4, 84, 84), dtype=np.uint8)).pin_memory().cuda()
            state=state/255.0
            hidden_state = (torch.rand((1, 64, 512), dtype=torch.float32).pin_memory().cuda(),
                            torch.rand((1, 64, 512), dtype=torch.float32).pin_memory().cuda())
            with torch.no_grad():
                _,hidden_state=model(state[:,0:burnin],hidden_state)

            x,hidden_state=model(state[:,burnin:],hidden_state)
            loss=torch.mean((x-target)**2)
        if autocast:
            scaler.scale(loss).backward()
            scaler.step(optimizer)
            scaler.update()
        else:
            loss.backward()
            optimizer.step()

def init_process(rank, size, fn, backend='gloo'):
    """ Initialize the distributed environment. """
    os.environ['MASTER_ADDR'] = '127.0.0.1'
    os.environ['MASTER_PORT'] = '29500'
    dist.init_process_group("nccl", rank=rank, world_size=size)
    fn(rank, size)


if __name__ == '__main__':
    size = 2
    processes = []
    mp.set_start_method("spawn")
    for rank in range(size):
        p = mp.Process(target=init_process, args=(rank, size, run))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()

```
WARNING: Logging before InitGoogleLogging() is written to STDERR
I20220618 20:37:05.296600 36344 ProcessGroupNCCL.cpp:588] [Rank 1] ProcessGroupNCCL initialized with following options:
NCCL_ASYNC_ERROR_HANDLING: 0
NCCL_DESYNC_DEBUG: 0
NCCL_BLOCKING_WAIT: 0
TIMEOUT(ms): 1800000
USE_HIGH_PRIORITY_STREAM: 0
NCCL_DEBUG: UNSET
I20220618 20:37:05.296622 36359 ProcessGroupNCCL.cpp:733] [Rank 1] NCCL watchdog thread started!
WARNING: Logging before InitGoogleLogging() is written to STDERR
I20220618 20:37:05.297032 36343 ProcessGroupNCCL.cpp:588] [Rank 0] ProcessGroupNCCL initialized with following options:
NCCL_ASYNC_ERROR_HANDLING: 0
NCCL_DESYNC_DEBUG: 0
NCCL_BLOCKING_WAIT: 0
TIMEOUT(ms): 1800000
USE_HIGH_PRIORITY_STREAM: 0
NCCL_DEBUG: UNSET
I20220618 20:37:05.297062 36361 ProcessGroupNCCL.cpp:733] [Rank 0] NCCL watchdog thread started!
I20220618 20:37:06.627071 36343 reducer.cpp:110] Reducer initialized with bucket_bytes_cap: 26214400 first_bucket_bytes_cap: 1048576
I20220618 20:37:06.627076 36344 reducer.cpp:110] Reducer initialized with bucket_bytes_cap: 26214400 first_bucket_bytes_cap: 1048576
I20220618 20:37:06.627182 36343 logger.cpp:218] [Rank 0]: DDP Initialized with: 
broadcast_buffers: 1
bucket_cap_bytes: 26214400
find_unused_parameters: 1
gradient_as_bucket_view: 0
has_sync_bn: 0
is_multi_device_module: 0
iteration: 0
num_parameter_tensors: 8
output_device: 0
rank: 0
total_parameter_size_bytes: 62046588
world_size: 2
backend_name: nccl
bucket_sizes: 4241468, 57805120
cuda_visible_devices: N/A
device_ids: 0
dtypes: float
initial_bucket_size_limits: 26214400, 1048576
master_addr: 127.0.0.1
master_port: 29500
module_name: R2D2
nccl_async_error_handling: N/A
nccl_blocking_wait: N/A
nccl_debug: N/A
nccl_ib_timeout: N/A
nccl_nthreads: N/A
nccl_socket_ifname: N/A
torch_distributed_debug: DETAIL
I20220618 20:37:06.627188 36344 logger.cpp:218] [Rank 1]: DDP Initialized with: 
broadcast_buffers: 1
bucket_cap_bytes: 26214400
find_unused_parameters: 1
gradient_as_bucket_view: 0
has_sync_bn: 0
is_multi_device_module: 0
iteration: 0
num_parameter_tensors: 8
output_device: 1
rank: 1
total_parameter_size_bytes: 62046588
world_size: 2
backend_name: nccl
bucket_sizes: 4241468, 57805120
cuda_visible_devices: N/A
device_ids: 1
dtypes: float
initial_bucket_size_limits: 26214400, 1048576
master_addr: 127.0.0.1
master_port: 29500
module_name: R2D2
nccl_async_error_handling: N/A
nccl_blocking_wait: N/A
nccl_debug: N/A
nccl_ib_timeout: N/A
nccl_nthreads: N/A
nccl_socket_ifname: N/A
torch_distributed_debug: DETAIL
I20220618 20:37:07.525991 36343 reducer.cpp:1271] [Rank 0]: Parameter head.weight at index 6 is marked as unused.
I20220618 20:37:07.526013 36343 reducer.cpp:1271] [Rank 0]: Parameter conv.bias at index 1 is marked as unused.
I20220618 20:37:07.526016 36343 reducer.cpp:1271] [Rank 0]: Parameter head.bias at index 7 is marked as unused.
I20220618 20:37:07.526019 36343 reducer.cpp:1271] [Rank 0]: Parameter conv.weight at index 0 is marked as unused.
I20220618 20:37:07.526011 36344 reducer.cpp:1271] [Rank 1]: Parameter head.bias at index 7 is marked as unused.
I20220618 20:37:07.526031 36344 reducer.cpp:1271] [Rank 1]: Parameter head.weight at index 6 is marked as unused.
I20220618 20:37:07.526036 36344 reducer.cpp:1271] [Rank 1]: Parameter conv.bias at index 1 is marked as unused.
I20220618 20:37:07.526037 36344 reducer.cpp:1271] [Rank 1]: Parameter conv.weight at index 0 is marked as unused.
I20220618 20:37:07.659709 36359 ProcessGroupNCCL.cpp:735] [Rank 1] NCCL watchdog thread terminated normally
I20220618 20:37:07.659713 36361 ProcessGroupNCCL.cpp:735] [Rank 0] NCCL watchdog thread terminated normally

### Versions

PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A
OS: Arch Linux (x86_64)
GCC version: (GCC) 12.1.0
Clang version: 13.0.1
CMake version: version 3.23.2
Python version: 3.10 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 11.7.64
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 2080 SUPER
GPU 1: NVIDIA GeForce RTX 2080 SUPER

Nvidia driver version: 515.48.07
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.4.0
/usr/lib/libcudnn_adv_infer.so.8.4.0
/usr/lib/libcudnn_adv_train.so.8.4.0
/usr/lib/libcudnn_cnn_infer.so.8.4.0
/usr/lib/libcudnn_cnn_train.so.8.4.0
/usr/lib/libcudnn_ops_infer.so.8.4.0
/usr/lib/libcudnn_ops_train.so.8.4.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Versions of relevant libraries:
[pip3] functorch==0.1.0
[pip3] numpy==1.22.4
[pip3] pytorch-forecasting==0.9.2
[pip3] pytorch-ignite==0.4.8
[pip3] torch==1.11.0
[pip3] torch-tb-profiler==0.3.1
[pip3] torchaudio==0.11.0+820b383
[pip3] torchfile==0.1.0
[pip3] torchmetrics==0.9.1
[pip3] torchtext==0.12.0a0+d7a34d6
[pip3] torchvision==0.14.0a0+c02d6ce
[conda] Could not collect

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501 @mcarilli @ptrblck

**Comment by ptrblck** [comment]
I think the behavior we are observing is similar to the one described [here](https://discuss.pytorch.org/t/autocast-and-torch-no-grad-unexpected-behaviour/93475/3?u=ptrblck).

In summary:
* `amp.autocast` uses an internal cache to avoid transforming parameters if the cached FP16 copies can be reused
* if the parameters are updated in each iteration, the cache is invalid, and the transformation to FP16 would be needed again
* in a `no_grad()` block, the parameters are not updated and thus the cached FP16 copies can be reused
* this seems to be happening here: the cached parameters are used, the original FP32 parameters are not used, and DDP raises the error and asks you to use `find_unused_parameters=True`
* the workaround could be to use `find_unused_parameters=True` or to use the suggested workflow in the [forum post](https://discuss.pytorch.org/t/autocast-and-torch-no-grad-unexpected-behaviour/93475/3?u=ptrblck)

Let me know if this description matches your observations.

**Comment by axeln31459** [comment]
Thanks ptrblck for pointing out the forum post, I missed that. Putting separate autocast blocks around the no-grad and grad blocks seems to work and find_unused_parameters=True as well. I'll try to look into it a bit more in terms of performance and what happens under the hood, but I will close it now.
Again, thanks to everyone for your help.
